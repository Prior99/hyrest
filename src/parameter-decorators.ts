import "reflect-metadata";

export interface BodyParameter {
    readonly index: number;
}

export function getBodyParameters(target: Object, propertyKey: string | symbol): BodyParameter[] {
    const bodyParameters = Reflect.getMetadata("api:route:bodyparameters", target, propertyKey);
    if (bodyParameters) {
        return bodyParameters;
    }
    const newBodyParameters: BodyParameter[] = [];
    Reflect.defineMetadata("api:route:bodyparameters", newBodyParameters, target, propertyKey);
    return newBodyParameters;
}

export function body(): ParameterDecorator {
    return (target: Object, propertyKey: string, index: number) => {
        const bodyParameters = getBodyParameters(target, propertyKey);
        bodyParameters.push({
            index,
        });
    };
}

export interface QueryParameter {
    readonly index: number;
    readonly name: string;
}

export function getQueryParameters(target: Object, propertyKey: string | symbol): QueryParameter[] {
    const queryParameters = Reflect.getMetadata("api:route:queryparameters", target, propertyKey);
    if (queryParameters) {
        return queryParameters;
    }
    const newQueryParameters: QueryParameter[] = [];
    Reflect.defineMetadata("api:route:queryparameters", newQueryParameters, target, propertyKey);
    return newQueryParameters;
}

export function query(name: string): ParameterDecorator {
    return (target: Object, propertyKey: string | symbol, index: number) => {
        const queryParameters = getQueryParameters(target, propertyKey);
        queryParameters.push({
            index,
            name,
        });
    };
}

export interface UrlParameter {
    readonly index: number;
    readonly name: string;
}

export function getUrlParameters(target: Object, propertyKey: string | symbol): UrlParameter[] {
    const urlParameters = Reflect.getMetadata("api:route:urlparameters", target, propertyKey);
    if (urlParameters) {
        return urlParameters;
    }
    const newUrlParameters: UrlParameter[] = [];
    Reflect.defineMetadata("api:route:urlparameters", newUrlParameters, target, propertyKey);
    return newUrlParameters;
}

export function param(name: string): ParameterDecorator {
    return (target: Object, propertyKey: string, index: number) => {
        const urlParameters = getUrlParameters(target, propertyKey);
        urlParameters.push({
            index,
            name,
        });
    };
}
