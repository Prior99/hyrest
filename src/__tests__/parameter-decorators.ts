import { body, getBodyParameters, param, getUrlParameters, query, getQueryParameters } from "../parameter-decorators";

test("@body()", () => {
    class TestController { //tslint:disable-line
        public method(@param("param1") param: string, @query("query1") query: string, @body() body: any) {}
    }

    const controller = new TestController();

    expect(getBodyParameters(controller, "method")).toMatchSnapshot();
    expect(getUrlParameters(controller, "method")).toMatchSnapshot();
    expect(getQueryParameters(controller, "method")).toMatchSnapshot();
});
