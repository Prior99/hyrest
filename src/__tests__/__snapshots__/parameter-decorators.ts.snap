// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@body() 1`] = `
Array [
  Object {
    "index": 2,
  },
]
`;

exports[`@body() 2`] = `
Array [
  Object {
    "index": 0,
    "name": "param1",
  },
]
`;

exports[`@body() 3`] = `
Array [
  Object {
    "index": 1,
    "name": "query1",
  },
]
`;
